netebpfext/net_ebpf_ext.c

```diff
@@ -638,27 +700,47 @@ net_ebpf_extension_initialize_wfp_components(_Inout_ void* device_object)
 
     NET_EBPF_EXT_LOG_ENTRY();
 
-    if (_fwp_engine_handle != NULL) {
+    if (_wfp_engine_handle != NULL) {
         // already registered
         goto Exit;
     }
 
-    session.flags = FWPM_SESSION_FLAG_DYNAMIC;
-
-    status = FwpmEngineOpen(NULL, RPC_C_AUTHN_WINNT, NULL, &session, &_fwp_engine_handle);
+    // It is critical that this specific session be a static session. We explicitly ensure this by making sure that the
+    // FWPM_PROVIDER::flags member is zero.
+    // This logic ensures that we do not break WFP's processing rules for sessions and transactions when we create
+    // other dynamic sessions later (during addition of filters).
+    RtlZeroMemory(&session, sizeof(FWPM_SESSION));
+    status = FwpmEngineOpen(NULL, RPC_C_AUTHN_WINNT, NULL, &session, &_wfp_engine_handle);
     NET_EBPF_EXT_BAIL_ON_API_FAILURE_STATUS(NET_EBPF_EXT_TRACELOG_KEYWORD_EXTENSION, "FwpmEngineOpen", status);
     is_engine_opened = TRUE;
 
-    status = FwpmTransactionBegin(_fwp_engine_handle, 0);
+    // Clean up stale WFP persisted state, if any.
+    _net_ebpf_extension_cleanup_state(_wfp_engine_handle);
```

Comment:
I was wrong about sessions being persistent by default. It is not. Please read [this](https://learn.microsoft.com/en-us/windows/win32/fwp/object-management#sessions). So I think we do not need to call cleanup_state in driver entry. Sorry about the back and forth.

Reply:
No worries.  Thanks for that update.  So would I be right in assuming that NetEbpfExt would hang forever at unload if it did not perform _any_ of these clean-up actions?  If not, we'd _still_ need to clean-up stale state in DriverEntry, right?

Reply:
To clarify I want you to try reverting back the global session to be dynamic and let the states be cleaned up automatically at driver unload. But keep the per-filter-context session as static. I _think_ that session can still be associated with the states created in the global dynamic session.

If that does not work, whatever you have put here will work.